File: genefunk/src/main/java/de/hip/pnp/genefunk/GeneFunkGenomeService.java
package de.hip.pnp.genefunk;

import org.springframework.stereotype.Service;

import javax.annotation.PostConstruct;
import javax.transaction.Transactional;
import java.util.Collection;
import java.util.List;

@Service
@Transactional
class GeneFunkGenomeService {

    final GeneFunkGenomeRepository repository;
    final GeneFunkGenomeFactory genomeFactory;

    public GeneFunkGenomeService(GeneFunkGenomeRepository repository, GeneFunkGenomeFactory genomeFactory) {
        this.repository = repository;
        this.genomeFactory = genomeFactory;
    }

    public void save(Collection<? extends GeneFunkGenome> genomes) {
        repository.saveAll(genomes);
    }

    public List<GeneFunkGenome> getAllGenomes() {
        return repository.findAll();
    }

    @PostConstruct
    private void populateData() {
        this.save(genomeFactory.initiateAllGenomes());
    }
}

File: src/main/kotlin/de/hipp/pnp/genefunk/GeneFunkGenomeService.kt
package de.hipp.pnp.genefunk

import de.hipp.pnp.base.entity.CharacterSpeciesEntity
import de.hipp.pnp.genefunk.GenefunkInfoProducer
import io.github.oshai.kotlinlogging.KotlinLogging
import jakarta.annotation.PostConstruct
import org.springframework.stereotype.Service

@Service
class GeneFunkGenomeService(
    val repository: GeneFunkGenomeRepository,
    val genefunkInfoProducer: GenefunkInfoProducer
) {

    private val log = KotlinLogging.logger {}

    @PostConstruct
    fun init() {
        val genomes: List<CharacterSpeciesEntity>? = genefunkInfoProducer.getAllSpecies()
        if (genomes.isNullOrEmpty()) {
            println("No genomes found for GeneFunk")
            return
        }
        val geneFunkGenomes = genomes.map { species ->
            GeneFunkGenome().apply {
                name = species.name
                description = species.description
                attributes = species.attributes.mapValues { entry -> entry.value.toInt() }.toMutableMap()
                features = species.features.map { it }.toMutableSet()
                genomeType = getGenomeType(species.name)
            }
        }
        println("genomes found for GeneFunk: ${geneFunkGenomes.size}")
        save(geneFunkGenomes)
    }

    private fun getGenomeType(name: String): GeneFunkGenomeType {
        when (name.lowercase()) {
            "mutts" -> return GeneFunkGenomeType.MUTTS
            "optimized" -> return GeneFunkGenomeType.OPTIMIZED
            "transhuman" -> return GeneFunkGenomeType.TRANSHUMAN
            else -> return GeneFunkGenomeType.ENGINEERED
        }
    }

    fun save(genomes: List<GeneFunkGenome>) {
        genomes.forEach {
            log.info { "Saving genome: ${it.name}" }
            if (repository.existsByName(it.name)) {
                log.warn { "Genome with name ${it.name} already exists, skipping save." }
                return@forEach
            }
            repository.saveAndFlush(it)
        }
    }

    fun allGenomes(): MutableList<GeneFunkGenome?> = repository.findAll()
}
